packages/protocol/contracts-0.8/common/EpochManager.sol

```diff
@@ -0,0 +1,322 @@
+    _transferOwnership(msg.sender);
+    setRegistry(registryAddress);
+    setEpochDuration(newEpochDuration);
+    carbonOffsettingPartner = _carbonOffsettingPartner;
+    epochManagerInitializer = _epochManagerInitializer;
+  }
+
+  // DESIGNDESICION(XXX): we assume that the first epoch on the L2 starts as soon as the system is initialized
+  // to minimize amount of "limbo blocks" the network should stop relatively close to an epoch number (but wigh enough time)
+  // to have time to call the function EpochInitializer.migrateEpochAndValidators()
+  function initializeSystem(
+    uint256 firstEpochNumber,
+    uint256 firstEpochBlock,
+    address[] memory firstElected
+  ) external onlyEpochManagerInitializer {
+    require(!systemAlreadyInitialized(), "Epoch system already initialized");
+    require(firstEpochNumber > 0, "First epoch number must be greater than 0");
+    require(firstEpochBlock > 0, "First epoch block must be greater than 0");
+    require(
+      firstEpochBlock <= block.number,
+      "First epoch block must be less or equal than current block"
+    );
+    require(firstElected.length > 0, "First elected validators must be greater than 0");
+    firstKnownEpoch = firstEpochNumber;
+    currentEpochNumber = firstEpochNumber;
+
+    Epoch storage _currentEpoch = epochs[currentEpochNumber];
+    _currentEpoch.firstBlock = firstEpochBlock;
+    _currentEpoch.startTimestamp = block.timestamp;
+
+    elected = firstElected;
+    epochManagerInitializer = address(0);
+  }
+
+  // TODO maybe "freezeEpochRewards" "prepareForNextEpoch"
+
+  /// start next epoch process.
+  /// it freezes the epochrewards at the time of execution,
+  /// and starts the distribution of the rewards.
+  function startNextEpochProcess() external nonReentrant {
+    require(systemAlreadyInitialized(), "Epoch system not initialized");
+    require(isTimeForNextEpoch(), "Epoch is not ready to start");
+    require(!isOnEpochProcess(), "Epoch process is already started");
+    epochProcessing.status = EpochProcessStatus.Started;
+
+    epochs[currentEpochNumber].rewardsBlock = block.number;
+
+    // calculate rewards
+    getEpochRewards().updateTargetVotingYield();
+
+    (
+      uint256 perValidatorReward,
+      uint256 totalRewardsVoter,
+      uint256 totalRewardsCommunity,
+      uint256 totalRewardsCarbonFund
+    ) = getEpochRewards().calculateTargetEpochRewards();
+
+    epochProcessing.perValidatorReward = perValidatorReward;
+    epochProcessing.totalRewardsVoter = totalRewardsVoter;
+    epochProcessing.totalRewardsCommunity = totalRewardsCommunity;
+    epochProcessing.totalRewardsCarbonFund = totalRewardsCarbonFund;
+
+    allocateValidatorsRewards();
+
+    emit EpochProcessingStarted(currentEpochNumber);
+  }
+
+  function finishNextEpochProcess(
+    address[] calldata groups,
+    address[] calldata lessers,
+    address[] calldata greaters
```

Comment:
An optimization to have less calldata is to use the index in a list of validators rather than passing the addresses. Would that make sense?

Reply:
we agreed to leave this for the end, as it will potentially take more gas